app/models/language/tag.rb

```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+class Language::Tag < ApplicationRecord
+  has_many :tag_subtags
+  has_and_belongs_to_many :subtags
+  belongs_to :preferred_value, class_name: 'Language::Tag', optional: true
+
+  before_validation :set_tag, :set_preferred_value
+
+  validates :tag, presence: true, uniqueness: true
+  validates_with Language::Validators::SubtagsTypeValidator
+  validates_with Language::Validators::ExtLangValidator
+
+  def first_subtag_of_type(type)
+    TagValue.first_subtag_of_type(subtags, type)
+  end
+
+  def suppressed_script
+    first_subtag_of_type('language').suppress_script
+  end
+
+  def script
+    script = first_subtag_of_type('script')
+    script unless script == suppressed_script
+  end
+
+  def region
+    first_subtag_of_type('region')
+  end
+
+  def extlang
+    first_subtag_of_type('extlang')
+  end
+
+  def tag_value(use_preferred = false)
+    TagValue.for_subtags(subtags, use_preferred)
+  rescue SubtagError => ex
+    self.errors.add :subtags, ex.message
+  end
+
+  def set_tag
```

Comment:
I was a little thrown off by method names like `set_tag` and `set_preferred_value` that look like setter methods, but don't take any arguments.  Changing isn't absolutely necessary, but just giving my thoughts as I go through the PR.

Reply:
I will admit I had flashbacks to talking about `set_X` and `assign_X` on the mezz. Definitely not wedded to the names.

Reply:
@barmintor and I talked this over.  We'll want to:

Rename methods to:
`set_tag_from_subtags!` and `set_preferred_tag_from_subtags!`

And also make them private, since there's no benefit to offering them to outside callers.

Reply:
This change to private methods was implemented, although it did require use of a `send` in one instance (also inside this class).